lib/src/dart.dart

```diff
@@ -174,3 +141,52 @@ class AnalyzerErrorGroup implements Exception {
   @override
   String toString() => errors.join('\n');
 }
+
+/// Precompiles the Dart executable at [executablePath] to a kernel file at
+/// [outputPath].
+///
+/// This file is also cached at [incrementalDillOutputPath] which is used to
+/// initialize the compiler on future runs.
+///
+/// The [packageConfigPath] should point at the package config file to be used
+/// for `package:` uri resolution.
+///
+/// The [name] is used to describe the executable in logs and error messages.
+Future<void> precompile({
+  @required String executablePath,
+  @required String incrementalDillOutputPath,
+  @required String name,
+  @required String outputPath,
+  @required String packageConfigPath,
+}) async {
+  ensureDir(p.dirname(outputPath));
+  ensureDir(p.dirname(incrementalDillOutputPath));
+  const platformDill = 'lib/_internal/vm_platform_strong.dill';
+  final sdkRoot = p.relative(p.dirname(p.dirname(Platform.resolvedExecutable)));
+  var client = await FrontendServerClient.start(
+    executablePath,
+    incrementalDillOutputPath,
+    platformDill,
+    sdkRoot: sdkRoot,
+    packagesJson: packageConfigPath,
+    printIncrementalDependencies: false,
+  );
+  try {
+    var result = await client.compile();
+
+    final highlightedName = log.bold(name);
+    if (result?.errorCount == 0) {
+      log.message('Precompiled $highlightedName.');
+      await File(incrementalDillOutputPath).copy(outputPath);
```

Comment:
So `dart run test:test` will create `.dart_tool/pub/incremental/test/test.dart.incremental.dill` and then that file will be copied to `.dart_tool/pub/bin/test/test.dart-x.y.z.snapshot` ?

I guess that's fine, we could also consider running the recompilation step whenever `.dart_tool/package_config.json` has a modification time newer than `.dart_tool/pub/incremental/test/test.dart-x.y.z.dill`.
Then we won't have to store things in `.dart_tool/pub/bin/` anymore.